File: app/lib/theme/gutter.dart
part of 'index.dart';

const _mini = 4.0;
const _small = 8.0;
const _normal = 12.0;
const _medium = 16.0;
const _large = 20.0;
const _xlarge = 24.0;

class _Inner {
  /// ---------[ 4.0 ] ---------
  late final EdgeInsets mini;

  /// ---------[ 8.0 ] ---------
  late final EdgeInsets small;

  /// ---------[ 12.0 ] ---------
  late final EdgeInsets normal;

  /// ---------[ 16.0 ] ---------
  late final EdgeInsets medium;

  /// ---------[ 20.0 ] ---------
  late final EdgeInsets large;

  /// ---------[ 24.0 ] ---------
  late final EdgeInsets xlarge;

  _Inner(EdgeInsets Function(double value) factor) {
    mini = factor(_mini);

    small = factor(_small);

    normal = factor(_normal);

    medium = factor(_medium);

    large = factor(_large);

    xlarge = factor(_xlarge);
  }
}

abstract class Gutter {
  static final all = _Inner((value) => EdgeInsets.all(value));

  static final horizontal = _Inner(
    (value) => EdgeInsets.symmetric(horizontal: value),
  );
  static final vertical = _Inner(
    (value) => EdgeInsets.symmetric(vertical: value),
  );

  static final left = _Inner((value) => EdgeInsets.only(left: value));
  static final right = _Inner((value) => EdgeInsets.only(right: value));
  static final top = _Inner((value) => EdgeInsets.only(top: value));
  static final bottom = _Inner((value) => EdgeInsets.only(bottom: value));

  static final bottomLess = _Inner(
    (value) => EdgeInsets.fromLTRB(value, value, value, 0),
  );
  static final topLess = _Inner(
    (value) => EdgeInsets.fromLTRB(value, 0, value, value),
  );
  static final leftLess = _Inner(
    (value) => EdgeInsets.fromLTRB(0, value, value, value),
  );
  static final rightLess = _Inner(
    (value) => EdgeInsets.fromLTRB(value, value, 0, value),
  );
}

File: victory/lib/styles/gutter.js
const path = require("path");
const fs = require("fs");
const output = path.join(__dirname, "_gutter.dart");

write();
function task(name, tempalte, alias) {
  return {
    name,
    func() {
      return `\tstatic const ${alias ?? name}_${4 * 1} =  ${tempalte(4 * 1)};
  static const ${alias ?? name}_${4 * 2} =  ${tempalte(4 * 2)};
  static const ${alias ?? name}_${4 * 3} =  ${tempalte(4 * 3)};
  static const ${alias ?? name}_${4 * 4} =  ${tempalte(4 * 4)};
  static const ${alias ?? name}_${4 * 5} =  ${tempalte(4 * 5)};
  static const ${alias ?? name}_${4 * 6} =  ${tempalte(4 * 6)};
  static const ${alias ?? name}_${4 * 7} =  ${tempalte(4 * 7)};
  static const ${alias ?? name}_${4 * 8} =  ${tempalte(4 * 8)};
  static const ${alias ?? name}_${4 * 9} =  ${tempalte(4 * 9)};
  static  const ${alias ?? name}_${4 * 10}  = ${tempalte(4 * 10)};
  `;
    },
  };
}

async function write(content) {
  const tasks = [
    task("all", (value) => `EdgeInsets.all(${value})`),
    task(
      "horizontal",
      (value) => `EdgeInsets.symmetric(horizontal: ${value})`,
      "x"
    ),
    task(
      "vertical",
      (value) => `EdgeInsets.symmetric(vertical: ${value})`,
      "y"
    ),
    task("left", (value) => `EdgeInsets.only(left: ${value})`),
    task("right", (value) => `EdgeInsets.only(right: ${value})`),
    task("top", (value) => `EdgeInsets.only(top: ${value})`),
    task("bottom", (value) => `EdgeInsets.only(bottom: ${value})`),
    task(
      "leftless",
      (value) => `EdgeInsets.fromLTRB(0, ${value}, ${value}, ${value})`
    ),
    task(
      "rightless",
      (value) => `EdgeInsets.fromLTRB(${value}, ${value}, 0, ${value})`
    ),
    task(
      "topless",
      (value) => `EdgeInsets.fromLTRB(${value}, 0, ${value}, ${value})`
    ),
    task(
      "bottomless",
      (value) => `EdgeInsets.fromLTRB(${value}, ${value}, ${value}, 0)`
    ),
  ];
  let c = `import 'package:flutter/material.dart';\n\n`;
  c += `abstract class AppGap {\n`;
  c += tasks.map((i) => i.func()).join("\n");
  c += "}";

  fs.writeFileSync(output, c, {
    encoding: "utf-8",
    flush: true,
  });
}
